// SpecSplit — Disaggregated Speculative Decoding
// gRPC service and message definitions for draft/target communication.

syntax = "proto3";

package specsplit;

option java_multiple_files = true;
option java_package = "com.specsplit.proto";

// =============================================================================
// Services
// =============================================================================

// DraftService — served by the Draft Worker (small LLM, cheap GPU).
// Generates speculative token trees from a given prompt context.
service DraftService {
  // Generate a speculative token tree of depth K from the current context.
  rpc GenerateDrafts (DraftRequest) returns (DraftResponse);

  // Health check / readiness probe.
  rpc Ping (PingRequest) returns (PingResponse);
}

// TargetService — served by the Target Worker (large LLM, expensive GPU).
// Verifies draft token trees using tree-attention and returns accepted tokens.
// Supports session-based KV caching to avoid prompt recomputation.
service TargetService {
  // Verify a draft token tree against the target model's distribution.
  // If session_id is provided, the server will reuse the KV cache from
  // prior calls in the same session and roll it back to the accepted prefix.
  rpc VerifyDrafts (VerifyRequest) returns (VerifyResponse);

  // Terminate a session and release its KV cache from GPU memory.
  rpc EndSession (EndSessionRequest) returns (EndSessionResponse);

  // Health check / readiness probe.
  rpc Ping (PingRequest) returns (PingResponse);
}

// =============================================================================
// Messages — Token Tree Structure
// =============================================================================

// A single node in the speculative token tree.
// Each node represents a candidate token with its draft probability.
message TokenNode {
  int32  token_id    = 1;  // Vocabulary index of the candidate token.
  float  log_prob    = 2;  // Log-probability assigned by the draft model.
  repeated TokenNode children = 3;  // Child candidates (branching factor).
  // Top-K draft distribution for correct residual sampling (covers ~99% mass).
  // When present, residual = max(0, P_target - P_draft) for full vocabulary.
  repeated int32 top_k_token_ids = 4;  // Token IDs in descending prob order.
  repeated float top_k_probs     = 5;  // Corresponding probabilities (same order).
}

// =============================================================================
// Messages — Draft Service
// =============================================================================

// Request to generate a speculative draft tree.
message DraftRequest {
  string  request_id     = 1;  // Unique request identifier for tracing.
  repeated int32 prompt_token_ids = 2;  // Tokenized prompt context.
  int32   max_draft_len  = 3;  // Maximum tree depth (K).
  int32   num_beams      = 4;  // Branching factor at each level.
  optional float temperature = 5;  // Sampling temperature (0 = greedy). Use HasField to distinguish unset from 0.
  bool    reset_cache    = 6;  // If true, clear draft KV cache before generating
                               // (used after speculation misses).
  string  session_id     = 7;  // Session ID for per-session KV cache isolation.
                               // Required for thread-safe concurrent requests.
}

// Response containing the generated draft tree.
message DraftResponse {
  string  request_id     = 1;  // Echo of the request identifier.
  repeated TokenNode draft_tree = 2;  // Root-level candidate tokens (forest).
  TelemetryMetadata telemetry  = 3;  // Server-side timing information.
}

// =============================================================================
// Messages — Target Service
// =============================================================================

// Request to verify a draft tree against the target model.
message VerifyRequest {
  string  request_id     = 1;  // Unique request identifier for tracing.
  repeated int32 prompt_token_ids = 2;  // Full prompt context (sent on round 0 or after cache flush).
  repeated TokenNode draft_tree    = 3;  // Draft tree to verify.
  string  session_id     = 4;  // Session ID for KV cache reuse across rounds.
                               // If empty, verification is stateless (no caching).
  float   temperature    = 5;  // Sampling temperature for verification.
                               // 0.0 = greedy, >0.0 = stochastic rejection sampling.
  repeated int32 new_token_ids = 6;  // Delta tokens since last verified position.
                                     // When non-empty, replaces prompt_token_ids for
                                     // context reconstruction (Issue 11 optimisation).
  int32 expected_prefix_length = 7;  // Orchestrator's prompt+generated length.
                                     // Target validates cache_state.seq_len == this;
                                     // if mismatch, force cache flush and rebuild.
}

// Response containing verification results.
message VerifyResponse {
  string  request_id        = 1;  // Echo of the request identifier.
  repeated int32 accepted_token_ids = 2;  // Longest accepted prefix from the tree.
  int32   correction_token_id = 3;  // Correction token sampled from target (if draft rejected).
  bool    has_correction      = 8;  // True if correction_token_id is valid (avoids token 0 conflation).
  int32   num_accepted       = 4;  // Number of draft tokens accepted.
  TelemetryMetadata telemetry = 5;  // Server-side timing information.
  string  session_id        = 6;  // Echo of session ID (confirms cache association).
  bool    cache_hit         = 7;  // True if an existing KV cache was reused.
}

// Request to end a session and release its KV cache.
message EndSessionRequest {
  string  session_id = 1;  // The session to terminate.
}

// Response confirming session termination.
message EndSessionResponse {
  string  session_id   = 1;  // Echo of the terminated session ID.
  bool    was_active   = 2;  // True if the session existed and was released.
}

// =============================================================================
// Messages — Telemetry & Utilities
// =============================================================================

// Timing and diagnostic metadata attached to RPC responses.
message TelemetryMetadata {
  string  span_id           = 1;  // Distributed trace span identifier.
  double  wall_time_ms      = 2;  // Wall-clock time for the RPC handler (ms).
  double  model_time_ms     = 3;  // Time spent in model forward pass (ms).
  int32   tokens_processed  = 4;  // Number of tokens processed in this call.
  string  device            = 5;  // Device identifier (e.g., "cuda:0").
}

// Simple health-check messages.
message PingRequest  {}
message PingResponse {
  string  status = 1;  // "ok" if healthy.
  string  worker_type = 2;  // "draft" or "target".
}
